code/app/src/main/java/com/kernelcrew/moodapp/ui/MoodComments.java

```diff
@@ -0,0 +1,303 @@
+import com.kernelcrew.moodapp.data.UserProvider;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Locale;
+
+
+/**
+ * Fragment responsible for displaying the comments on a selected mood.
+ *
+ */
+public class MoodComments extends Fragment implements MoodHistoryAdapter.OnItemClickListener {
+    /** RecyclerView for displaying mood comment items */
+    private RecyclerView recyclerView;
+    NavigationBarView navigationBar;
+    private BottomNavBarController navBarController;
+    private ImageView imageMoodIcon;
+    private TextView tvMoodState, usernameText, dayTimeText, commentCountText;
+
+    /** Adapter for binding mood data to the RecyclerView */
+    private MoodEventProvider provider;
+    // Add CommentProvider
+    private CommentProvider commentProvider;
+    private UserProvider userProvider;
+    /** List to store mood events */
+    private List<MoodEvent> moods = new ArrayList<>();
+    /** Registration for Firestore listener */
+    private ListenerRegistration snapshotListener;
+    private FirebaseFirestore db;
+    private String moodEventId;
+    private String sourceScreen;
+    private String userId;
+    private String username;
+
+    // Add these variables for the input fields
+    private TextInputLayout searchInputLayout;
+    private TextInputEditText searchInput;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        // Retrieve arguments passed from previous screen
+        if (getArguments() != null) {
+            moodEventId = getArguments().getString("moodEventId");
+            sourceScreen = getArguments().getString("sourceScreen", "home");
+        }
+        // Initialize Firestore
+        db = FirebaseFirestore.getInstance();
+        provider = MoodEventProvider.getInstance();
+        // Initialize UserProvider
+        userProvider = UserProvider.getInstance();
+        // Initialize CommentProvider
+        commentProvider = CommentProvider.getInstance();
+        // Get current user ID
+        FirebaseUser currentUser = FirebaseAuth.getInstance().getCurrentUser();
+        if (currentUser != null) {
+            userId = currentUser.getUid();
+            userProvider.getUsername(userId)
+                    .addOnSuccessListener(result -> {
+                        username = result;
+                    })
+                    .addOnFailureListener(e -> {
+                        Log.e("MoodComments", "Error fetching username", e);
+                    });
+        }
+    }
+
+    /**
+     * Creates and returns the view hierarchy associated with the fragment.
+     */
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_mood_comments, container, false);
+
+        provider = MoodEventProvider.getInstance();
```

Comment:
We also assign `provider` an `userProvider` in `onCreate`. We should only need to do that once.